internal/plugin/service/organization/org/organization_data_source.go

```diff
@@ -136,69 +103,72 @@ func (r *organizationDataSource) ConfigValidators(_ context.Context) []datasourc
 	}
 }
 
-// fillModel fills the organization data source model from the Aiven API.
-func (r *organizationDataSource) fillModel(ctx context.Context, model *organizationDataSourceModel) (err error) {
-	normalizedID, err := schemautil.NormalizeOrganizationID(ctx, r.client, model.ID.ValueString())
-	if err != nil {
-		return
-	}
-
-	account, err := r.client.Accounts.Get(ctx, normalizedID)
-	if err != nil {
-		return
-	}
-
-	model.Name = types.StringValue(account.Account.Name)
-
-	model.TenantID = types.StringValue(account.Account.TenantId)
-
-	model.CreateTime = types.StringValue(account.Account.CreateTime.String())
-
-	model.UpdateTime = types.StringValue(account.Account.UpdateTime.String())
-
-	return
-}
-
 // Read reads an organization data source.
 func (r *organizationDataSource) Read(ctx context.Context, req datasource.ReadRequest, resp *datasource.ReadResponse) {
-	var state organizationDataSourceModel
-
-	resp.Diagnostics.Append(req.Config.Get(ctx, &state)...)
+	var config organizationDataSourceModel
+	resp.Diagnostics.Append(req.Config.Get(ctx, &config)...)
 	if resp.Diagnostics.HasError() {
 		return
 	}
 
-	if state.ID.IsNull() {
-		list, err := r.client.Accounts.List(ctx)
-		if err != nil {
-			resp.Diagnostics = util.DiagErrorReadingDataSource(resp.Diagnostics, r, err)
+	// If ID is not set, find by name
+	var organizationID string
+	if !config.ID.IsNull() {
+		organizationID = config.ID.ValueString()
 
+		// Resolve the organization ID to account ID if provided
+		var err error
+		organizationID, err = ResolveAccountID(ctx, r.client, organizationID)
+		if err != nil {
+			r.diag.AddError(&resp.Diagnostics, "resolving account ID", err)
+			return
+		}
+	} else if !config.Name.IsNull() {
+		// List organizations and find by name
+		accounts, err := r.client.AccountList(ctx)
+		if err != nil {
+			r.diag.AddError(&resp.Diagnostics, "reading", err)
 			return
 		}
 
 		var found bool
-
-		for _, account := range list.Accounts {
-			if account.Name == state.Name.ValueString() {
+		for _, account := range accounts {
+			if account.AccountName == config.Name.ValueString() {
 				if found {
-					resp.Diagnostics = util.DiagDuplicateFoundByName(resp.Diagnostics, state.Name.ValueString())
-
+					resp.Diagnostics.AddError(
+						"Multiple organizations found with the same name",
+						fmt.Sprintf("Multiple organizations found with name %s, please use ID instead", config.Name.ValueString()),
+					)
 					return
 				}
-
-				state.ID = types.StringValue(account.OrganizationId)
-
+				organizationID = account.AccountId
 				found = true
 			}
 		}
+
+		if !found {
+			resp.Diagnostics.AddError(
+				"Organization not found",
+				fmt.Sprintf("No organization found with name %s", config.Name.ValueString()),
+			)
+			return
+		}
 	}
```

Comment:
It either should have `else` statement or must validate the `organizationID` not empty.
UPD: at least one of id or name is required https://github.com/aiven/terraform-provider-aiven/blob/2b96cbf0a2181bf53f2b664a41e90458bed9e585/internal/plugin/service/organization/org/organization_data_source.go#L102
Then we could turn `else if` into `else` and add a ref to the validation, or validate the `organizationID` eventually. I personally would go for the validation, as it more solid/durable solution.

Reply:
Hey, could you elaborate on how to do the validation?

Reply:
Ah, sure:
```go
// It's empty
var organizationID string

// Tries to find a value
switch {
case !config.ID.IsNull():
    ...
case !config.Name.IsNull():
    ...
}

// Secures the rest of the code
if organizationID == "" {
    return fmt.Errorf(...)
}
```

`organizationID == ""` secures from issues in the `switch` that might happen in future.

Reply:
Got it thanks, implemented in https://github.com/aiven/terraform-provider-aiven/pull/2117/commits/afb2558bfd924930fdb154170fd522ba0997cd95